longitudinal_controller/src/longitudinal_controller.py

```diff
@@ -0,0 +1,227 @@
+        """
+        Callback function for current velocity message
+        """
+        self.current_velocity = msg.twist.linear.x
+
+    def firmware_log_callback(self, msg: ArduinoLogging):
+        """
+        Callback of log message from low level controller.
+        """
+        self.speed_controller_enabled = msg.engaged_mode
+        self.time_last_enable_msg = time.time()
+        # TODO: Add timer for safety?
+
+    def target_velocity_callback(self, msg: TwistStamped):
+        """
+        Callback function for target velocity message. Main function of the controller.
+        """
+        self.desired_velocity = msg.twist.linear.x
+
+    def loop(self):
+        """
+        Main loop of the longitudinal controller
+        """
+
+        while not rospy.is_shutdown():
+
+            time_since_last_message = time.time() - self.time_last_enable_msg
+
+            if self.longitudinal_control_mode == LongitudinalControlMode.PID.value:
+
+                if (
+                    self.speed_controller_enabled
+                    and time_since_last_message
+                    < self.timeout_engage_msg_before_stop_secs
+                ):
+                    rospy.logwarn("-------------")
+                    rospy.logwarn("Controller ON")
+
+                    # Compute sample time
+                    dt = time.time() - self.previous_t
+
+                    # Compute velocity error
+                    error = self.desired_velocity - self.current_velocity
+                    rospy.logwarn("Error: " + str(error))
+
+                    # Call the PID controller
+                    throttle_diff, p, i, d = self.pid_controller.update(
+                        error, self.anti_windup_enabled, dt
+                    )
+                    rospy.logwarn("Throttle diff: " + str(throttle_diff))
+
+                    # Add the controller's output value to the idle position
+                    throttle_value = self.throttle_idle_autonomous_pwm + throttle_diff
+
+                    #  --- Anti windup strategy ---
+
+                    # Check if throttle output is already saturating the actuator
+                    if (
+                        throttle_value < self.throttle_min_autonomous_pwm
+                        or throttle_value > self.throttle_max_autonomous_pwm
+                    ):
+                        throttle_saturating = True
+
+                        if throttle_value < self.throttle_min_autonomous_pwm:
+                            throttle_value = self.throttle_min_autonomous_pwm
+
+                        elif throttle_value > self.throttle_max_autonomous_pwm:
+                            throttle_value = self.throttle_max_autonomous_pwm
+
+                    else:
+                        throttle_saturating = False
+
+                    # Check if controller is trying to saturate the actuator even more
+                    if np.sign(error) == np.sign(throttle_value):
```

Comment:
Shouldn't this be:
```suggestion
                    if np.sign(error) == np.sign(throttle_diff):
```
?

Reply:
Good catch. Indeed.